src/sync/stream.ts

```diff
@@ -61,6 +74,147 @@ function isStreamSyncHistoryError(data: IStreamSyncHistoryData): data is IStream
   return data.type === 'stream:history:error';
 }
 
+interface IStreamItemVertex {
+  seq: number;
+  type: 'vertex';
+  vertex: IHistoryTx;
+}
+
+interface IStreamItemAddress {
+  seq: number;
+  type: 'address';
+  address: IAddressInfo;
+}
+
+type IStreamItem = IStreamItemAddress | IStreamItemVertex;
+
+function isStreamItemVertex(item: IStreamItem): item is IStreamItemVertex {
+  return item.type === 'vertex';
+}
+
+function isStreamItemAddress(item: IStreamItem): item is IStreamItemAddress {
+  return item.type === 'address';
+}
+
+/**
+ * Stream statistics manager.
+ * Will provide insight on the rates of the stream and how the client is performing.
+ */
+class StreamStatsManager {
+  // Counter for received events
+  recvCounter: number;
+
+  // Counter for processed events
+  procCounter: number;
+
+  // Counter for ack messages sent
+  ackCounter: number;
+
+  // Counter for the number of times the queue has been empty.
+  emptyCounter: number;
+
+  // Timer to process in_rate, which is number of received events per second.
+  inTimer?: ReturnType<typeof setTimeout>;
+
+  // Timer to process out_rate, which is the number of processed events per second.
+  outTimer?: ReturnType<typeof setTimeout>;
+
+  // Simple timer to log the number of events on queue.
+  qTimer?: ReturnType<typeof setTimeout>;
+
+  // Custom logger to log in the desired format.
+  logger: ILogger;
+
+  // Reference to the item queue we are processing.
+  q: Queue;
+
+  // Interval to use when calculating rates.
+  dt: number;
+
+  constructor(queue: Queue, logger: ILogger) {
+    this.recvCounter = 0;
+    this.procCounter = 0;
+    this.ackCounter = 0;
+    this.emptyCounter = 0;
+
+    this.logger = logger;
+    this.dt = 2000;
+    this.q = queue;
+    this.qTimer = setInterval(() => {
+      this.logger.debug(`[*] queue_size: ${queue.size()}`);
+    }, this.dt);
+  }
+
+  /**
+   * Perform any cleanup necessary when the queue is done processing.
+   * For instance, stop the timers processing the io rates.
+   */
+  clean() {
+    if (this.inTimer) {
+      clearTimeout(this.inTimer);
+    }
+    if (this.outTimer) {
+      clearTimeout(this.outTimer);
+    }
+    if (this.qTimer) {
+      clearInterval(this.qTimer);
+    }
+  }
+
+  /**
+   * Mark an ACK message sent to the fullnode.
+   */
+  ack(seq: number) {
+    this.ackCounter += 1;
+    this.logger.debug(
+      `[*] ACKed ${seq} with queue at ${this.q.size()}. Sent ACK ${this.ackCounter} times`
+    );
+  }
+
+  /**
+   * Mark that an item has been received by the queue.
+   * This also manages the inTimer, which calculates the rate of received items.
+   */
+  recv() {
+    if (!this.inTimer) {
+      this.inTimer = setTimeout(() => {
+        this.logger.debug(`[+] => in_rate: ${(1000 * this.recvCounter) / this.dt} items/s`);
```

Comment:
The current method of calculating `in_rate` will not be really useful because it provides an average over the entire duration of the stream, from start to finish. Instead, we should calculate the average over the last N seconds (with N being between 5 and 10 seconds) to get a more accurate, real-time reflection of the rate. To further improve this, we can smooth the result using an Exponential Moving Average (EMA).

Reply:
You are already sampling every 2s. Thanks for clarifying.